runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -161,19 +166,43 @@ local function diagnostic_vim_to_lsp(diagnostics)
 end
 
 ---@type table<integer,integer>
-local _client_namespaces = {}
+local _client_push_namespaces = {}
+---@type table<integer,integer>
+local _client_pull_namespaces = {}
 
---- Get the diagnostic namespace associated with an LSP client |vim.diagnostic|.
+--- Get the diagnostic namespace associated with an LSP client |vim.diagnostic| for diagnostics
 ---
 ---@param client_id integer The id of the LSP client
-function M.get_namespace(client_id)
+---@param is_pull boolean Whether the namespace is for a pull or push client
+function M.get_namespace(client_id, is_pull)
   vim.validate({ client_id = { client_id, 'n' } })
-  if not _client_namespaces[client_id] then
-    local client = vim.lsp.get_client_by_id(client_id)
-    local name = string.format('vim.lsp.%s.%d', client and client.name or 'unknown', client_id)
-    _client_namespaces[client_id] = vim.api.nvim_create_namespace(name)
+
+  local namespace_table
+  local key
+  local name
+  local client = vim.lsp.get_client_by_id(client_id)
+
+  if is_pull then
+    namespace_table = _client_pull_namespaces
+    local server_id = vim.tbl_get(client.server_capabilities, 'diagnosticProvider', 'identifier')
```

Comment:
I think if the server uses dynamic capabilities, this won't be within the `server_capabilities.diagnosticProvider` but instead will be part of `client.dynamic_capabilities`. See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticRegistrationOptions - it extends [DiagnosticOptions](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticOptions)


---


I also wonder if we should expose the namespaces via the `lsp.Client` object.
E.g. it could have `namespaces` table has has optional keys per module/feature.

`client.namespaces.diagnostic.{push,pull}`
or `client.namespaces.diagnostic_{push,pull}`, not sure how much nesting we want.

We'd then move away from the `vim.lsp.with` configuration approach, and instead could tell users to use `vim.diagnostic.config(opts, client.namespaces.diagnostic.pull)` in an `LspAttach` autocmd callback.


But then we'll also have to figure out a way to deal with capability changes via the dynamic registration. (Maybe we have to introduce another `LspCapability` event, that is fired _per_ capability during the initialization to normalize between dynamic capability registration, and static registration, but that's a topic. But that's related to https://github.com/neovim/neovim/issues/24229)


(Please wait for other inputs before acting on this)

Reply:
>I also wonder if we should expose the namespaces via the lsp.Client object.
E.g. it could have namespaces table has has optional keys per module/feature.

I like this, and strongly prefer this to the existing `vim.lsp.diagnostic.get_namespace()`.

Reply:
> I think if the server uses dynamic capabilities, this won't be within the `server_capabilities.diagnosticProvider` but instead will be part of `client.dynamic_capabilities`. See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticRegistrationOptions - it extends [DiagnosticOptions](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticOptions)

Hmmm....I had some problems understanding this part of the spec, and wasn't able to figure out how to test it.

The way I understood this was that `DiagnosticOptions` is the type of the `diagnosticProvider` field of `ServerCapabilities` here - https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#serverCapabilities

Reply:
Indeed, when I inspect `client.server_capabilities` here, I do see `diagnosticProvider`:
```lua
client.server_capabilities = {
  ...
  diagnosticProvider = { interFileDependencies = false, workspaceDiagnostics = false },
  ...
}
```

Reply:
> > I also wonder if we should expose the namespaces via the lsp.Client object.
> > E.g. it could have namespaces table has has optional keys per module/feature.
> 
> I like this, and strongly prefer this to the existing `vim.lsp.diagnostic.get_namespace()`.

Can we do this in a separate PR?

Reply:
> Can we do this in a separate PR?

Yes that's okay

> Indeed, when I inspect client.server_capabilities here, I do see diagnosticProvider:

Yes, but it could instead also be within the dynamic capabilities if the server is using them instead of static capabilities.
We can also set `dynamicRegistration = false` in the capability for now, and follow up on this later.